src/core/generalized_first_order.jl

```diff
@@ -0,0 +1,279 @@
+
+function SciMLBase.__init(prob::AbstractNonlinearProblem{uType, iip},
+        alg::GeneralizedFirstOrderAlgorithm, args...; alias_u0 = false, maxiters = 1000,
+        abstol = nothing, reltol = nothing, maxtime = nothing,
+        termination_condition = nothing, internalnorm = DEFAULT_NORM, linsolve_kwargs = (;),
+        kwargs...) where {uType, iip}
+    timer = get_timer_output()
+    @static_timeit timer "cache construction" begin
+        (; f, u0, p) = prob
+        u = __maybe_unaliased(u0, alias_u0)
+        fu = evaluate_f(prob, u)
+        @bb u_cache = copy(u)
+
+        linsolve = get_linear_solver(alg.descent)
+
+        abstol, reltol, termination_cache = init_termination_cache(abstol, reltol, fu, u,
+            termination_condition)
+        linsolve_kwargs = merge((; abstol, reltol), linsolve_kwargs)
+
+        jac_cache = JacobianCache(prob, alg, f, fu, u, p; autodiff = alg.jacobian_ad,
+            linsolve, jvp_autodiff = alg.forward_ad, vjp_autodiff = alg.reverse_ad)
+        J = jac_cache(nothing)
+        descent_cache = SciMLBase.init(prob, alg.descent, J, fu, u; abstol, reltol,
+            internalnorm, linsolve_kwargs, timer)
+        du = get_du(descent_cache)
+
+        if alg.trustregion !== missing && alg.linesearch !== missing
+            error("TrustRegion and LineSearch methods are algorithmically incompatible.")
+        end
+
+        GB = :None
+        linesearch_cache = nothing
+        trustregion_cache = nothing
+
+        if alg.trustregion !== missing
+            supports_trust_region(alg.descent) || error("Trust Region not supported by \
+                                                        $(alg.descent).")
+            trustregion_cache = init(prob, alg.trustregion, f, fu, u, p; internalnorm,
+                kwargs...)
+            GB = :TrustRegion
+        end
+
+        if alg.linesearch !== missing
+            supports_line_search(alg.descent) || error("Line Search not supported by \
+                                                        $(alg.descent).")
+            linesearch_cache = init(prob, alg.linesearch, f, fu, u, p; internalnorm,
+                kwargs...)
+            GB = :LineSearch
+        end
+
+        trace = init_nonlinearsolve_trace(alg, u, fu, ApplyArray(__zero, J), du; kwargs...)
+
+        return GeneralizedFirstOrderAlgorithmCache{iip, GB, maxtime !== nothing}(fu, u,
+            u_cache, p, du, J, alg, prob, jac_cache, descent_cache, linesearch_cache,
+            trustregion_cache, 0, 0, maxiters, maxtime, alg.max_shrink_times, timer, 0.0,
+            true, termination_cache, trace, ReturnCode.Default, false)
+    end
+end
+
+function __step!(cache::GeneralizedFirstOrderAlgorithmCache{iip, GB};
+        recompute_jacobian::Union{Nothing, Bool} = nothing, kwargs...) where {iip, GB}
+    @static_timeit cache.timer "jacobian" begin
+        if (recompute_jacobian === nothing || recompute_jacobian) && cache.make_new_jacobian
```

Comment:
So what's the plan on the ODE solver for triggering new Jacobians? I didn't see the high level function that would be involved. I assume it's not just flipping `make_new_jacobian`?

Reply:
There are 2 possibilities:

1. the ODE solver uses the iterator form, and passes in `recompute_jacobian = false` if it wants to prevent a new jacobian construction
2. we have an inbuilt SingleFactorizeNewton which is similar to KINSOL and sets the make_new_jacobian internally.

Reply:
> the ODE solver uses the iterator form, and passes in recompute_jacobian = false if it wants to prevent a new jacobian construction

Oh I missed that. That works.